### PR TITLE
persist-txn: Use efficient timestamp rewrites

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, if persist-txn failed to commit a transaction, then it
would delete the batch associated with the failed transaction and
create a new batch for every commit retry. Deleting and creating
batches are potentially expensive operations.

In 1f861336ce368fffecfdb11c8277816243257d4a, an efficient method was
added to rewrite the timestamp of an existing batch. This commit uses
that new method in persist-txn to rewrite the timestamp of a batch when
a commit fails instead of deleting and creating a new batch.

Part of #22173

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/persist-txn/src/txns.rs [hunk 1]
```diff
@@ -707,6 +722,23 @@ where
     }
 }
 
+#[cfg(test)]
+impl<K, V, T, D, O, C> TxnsHandle<K, V, T, D, O, C>
```
Comment (danhhz) [inline src/persist-txn/src/txns.rs:726]: super nit: can we put this next to the definition of `TestTxn` (inside the big `#[cfg(test)]mod tests` block) instead?

I have an irrational dislike for `#[cfg(test)]` gated `use` imports at the top of the file :)

**Review by danhhz — APPROVED** [review]
Looks fantastic!

You'd mentioned potentially adding a targeted unit test for this, did the existing ones end up having enough coverage already?

As we discussed offline, the only thing other than unit tests that would cover this atm is the persist-txn maelstrom tests, so I'll kick one of those off

**Comment by jkosh44** [comment]
> You'd mentioned potentially adding a targeted unit test for this, did the existing ones end up having enough coverage already?
>
> As we discussed offline, the only thing other than unit tests that would cover this atm is the persist-txn maelstrom tests, so I'll kick one of those off

Oops, this completely slipped my mind, I'll add one. I think the scenario that the existing tests probably don't cover is:
1. Add writes to txn.
2. Commit and fail.
3. Add more writes to txn.
4. Commit and succeed.

File: src/persist-txn/src/txns.rs (hunk 1 above)
Comment (jkosh44) [reply src/persist-txn/src/txns.rs]: Done

**Pull request merged**